src_c/time.c

```diff
@@ -28,76 +28,140 @@
+typedef struct pgEventTimer {
+    struct pgEventTimer *next;
+    pgEventObject *event;
+    int repeat;
+} pgEventTimer;
 
-static SDL_TimerID event_timers[pgNUMEVENTS] = {0};
+static pgEventTimer *pg_event_timer = NULL;
+static SDL_mutex *timermutex = NULL;
 
-#if IS_SDLv2
-static size_t
-enumerate_event(Uint32 type)
+static int
+_pg_add_event_timer(pgEventObject *ev, int repeat)
 {
-    assert(pgNUMEVENTS == 1 + 15 + (PG_NUMEVENTS - PGE_USEREVENT));
-    switch (type) {
-        case SDL_ACTIVEEVENT:
-            return 1;
-        case SDL_KEYDOWN:
-            return 2;
-        case SDL_KEYUP:
-            return 3;
-        case SDL_MOUSEMOTION:
-            return 4;
-        case SDL_MOUSEBUTTONDOWN:
-            return 5;
-        case SDL_MOUSEBUTTONUP:
-            return 6;
-        case SDL_JOYAXISMOTION:
-            return 7;
-        case SDL_JOYBALLMOTION:
-            return 8;
-        case SDL_JOYHATMOTION:
-            return 9;
-        case SDL_JOYBUTTONDOWN:
-            return 10;
-        case SDL_JOYBUTTONUP:
-            return 11;
-        case SDL_VIDEORESIZE:
-            return 12;
-        case SDL_VIDEOEXPOSE:
-            return 13;
-        case SDL_QUIT:
-            return 14;
-        case SDL_SYSWMEVENT:
-            return 15;
+    pgEventTimer *new;
+    SDL_LockMutex(timermutex);
+
+    new = PyMem_New(pgEventTimer, 1);
+    if (!new) {
+        SDL_UnlockMutex(timermutex);
+        return 0;
     }
-    if (type >= PGE_USEREVENT && type < PG_NUMEVENTS)
-        return type - PGE_USEREVENT + 16;
-    return 0;
+
+    new->next = pg_event_timer;
+    new->event = ev;
+    new->repeat = repeat;
+    pg_event_timer = new;
+
+    SDL_UnlockMutex(timermutex);
+    return 1;
 }
-#endif /* IS_SDLv2 */
 
+static void
+_pg_remove_event_timer(int type)
+{
+    pgEventTimer *hunt, *prev = NULL;
+
+    SDL_LockMutex(timermutex);
+    if (pg_event_timer) {
+        hunt = pg_event_timer;
+        while (hunt->event->type != type) {
+            prev = hunt;
+            hunt = hunt->next;
+            if (!hunt)
+                break;
+        }
+        if (hunt) {
+            if (prev)
+                prev->next = hunt->next;
+            else
+                pg_event_timer = hunt->next;
+            Py_DECREF(hunt->event);
+            PyMem_Del(hunt);
+        }
+    }
+    SDL_UnlockMutex(timermutex);
+}
 
-static Uint32
-timer_callback(Uint32 interval, void *param)
+static pgEventTimer *
+_pg_get_event_on_timer(int type)
 {
-    if (SDL_WasInit(SDL_INIT_VIDEO)) {
-        SDL_Event event;
-        memset(&event, 0, sizeof(event));
-        event.type = (intptr_t)param;
-        SDL_PushEvent(&event);
+    pgEventTimer *hunt;
+
+    SDL_LockMutex(timermutex);
+
+    if (pg_event_timer) {
+        hunt = pg_event_timer;
+        do {
+            if (hunt->event->type == type) {
+                if (hunt->repeat >= 0)
+                    hunt->repeat--;
+                SDL_UnlockMutex(timermutex);
```

Comment:
Better to do cleanup and return in one place at the end.

Reply:
Will do :)

Reply:
done :)